File: 476_Number_Complement.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/number-complement/
# Given a positive integer, output its complement number. The complement strategy is to flip the bits of its
# binary representation.

# Find the number with all 1s that is the same length as num. Then take the XOR to flip the bits.
# Time - O(log n)
# Space - O(1)

class Solution(object):
    def findComplement(self, num):
        """
        :type num: int
        :rtype: int
        """
        i = 1

        while i <= num:
            i <<= 1

        return (i - 1) ^ num

File: java/476_Number_Complement.java
/*
https://leetcode.com/problems/number-complement/
Given a positive integer, output its complement number. The complement strategy is to flip the bits of its binary representation.
The given integer is guaranteed to fit within the range of a 32-bit signed integer.
You could assume no leading zero bit in the integer’s binary representation.

highestOneBit(num) is the number with only the highest bit of num set. Subtract 1 to set all lower bits. This creates a
mask. AND of mask with ~num sets all bits that were not set in num.
Time - O(log n)
Space - O(1)
*/

public class Solution {
    public int findComplement(int num) {

        return ~num & (Integer.highestOneBit(num) - 1);

    }
}
